### DISCUSSION TITLE
Failed to collect enough transaction endorsements

### DISCUSSION THREAD

**Discussion by yizhihuaqi** [hyperledger/fabric#4741] [Q&A]
I had designed the collections_config.json file. When I invoke my smart contract's function 'CreateDataSet' in a go application, use 'PutPrivateData' to put my dataset into a collection, I sometimes get the 'Failed to collect enough transaction endorsements ' erorr. But when I invote my 'CreateDataSet' function through the terminal, I never receive such erorr. What's wrong with my program? Hope to receive your help.

**Comment by jehanxaibahmed** [comment]
This is almost always a peer-targeting / private-data-collection mismatch between your CLI test and your Go client, not a bug in CreateDataSet itself.

When you invoke via the `peer chaincode invoke` CLI, you pass explicit `--peerAddresses` flags, so you fully control which peers endorse — typically the peers that are actually members of the private data collection.

When you invoke the same function from a Fabric SDK/Gateway client, the SDK usually relies on service discovery to pick endorsing peers, and discovery doesn't automatically know that `CreateDataSet` writes to a specific private data collection. If discovery selects peers from orgs that aren't listed in that collection's `collections_config.json` `policy`, those peers can't endorse the private-data write, and you fall short of the collection's endorsement policy — which surfaces as "failed to collect enough transaction endorsements" rather than a more specific private-data error.

Things I would check, in order. First, in your Go client, are you specifying endorsing peers explicitly (e.g. Gateway's `WithEndorsingOrganizations`), matching the orgs in the collection's `policy`? If discovery is picking automatically, that's the likely culprit. Second, confirm every peer in the collection's member list has the private-data-enabled chaincode installed and isn't lagging on the collection config. Third, check `requiredPeerCount` / `maxPeerCount` in `collections_config.json` — if `requiredPeerCount` exceeds the discoverable eligible peers, dissemination fails before endorsement completes. Fourth, enable debug logging on the endorsing peers and look for `ENDORSEMENT_POLICY_FAILURE` vs a private-data dissemination timeout — they point to different fixes. Fifth, some older SDK discovery clients don't filter peers by collection membership correctly; pinning explicit peers sidesteps this and is the standard workaround.

If you can share which client API you're calling `CreateDataSet` through (Gateway, fabric-sdk-go, or raw gRPC) and your `collections_config.json`, I can be more specific about where the mismatch is.
